src/codec.ts

```diff
@@ -231,6 +252,76 @@ export class PGNumeric {
   }
 }
 
+/**
+ * @typedef ProtoMessage
+ * @see Spanner.protoMessage
+ */
+export class ProtoMessage {
+  value: Buffer;
+  fullName: string;
+  messageFunction?: Function;
+
+  constructor(protoMessageParams: IProtoMessageParams) {
+    this.fullName = protoMessageParams.fullName;
+
+    if (protoMessageParams.value instanceof Buffer) {
+      this.value = protoMessageParams.value;
+    } else if (protoMessageParams.messageFunction) {
+      this.value = protoMessageParams.messageFunction['encode'](
+        protoMessageParams.value
+      ).finish();
+    } else {
+      throw new GoogleError(`protoMessageParams cannot be used for constructing ProtoMessage. 
+          Please pass serializedBytes as value or message object with messageFunction`);
+    }
+
+    this.messageFunction = protoMessageParams.messageFunction;
+  }
+
+  toJSON(): string {
+    if (this.messageFunction) {
+      return this.messageFunction['toObject'](
+        this.messageFunction['decode'](this.value)
+      );
+    }
+    return this.value.toString();
+  }
+}
+
+/**
+ * @typedef ProtoEnum
+ * @see Spanner.protoEnum
+ */
+export class ProtoEnum {
+  value: string;
+  fullName: string;
+  enumObject?: object;
+
+  constructor(protoEnumParams: IProtoEnumParams) {
+    this.fullName = protoEnumParams.fullName;
+
+    if (/^\d+$/.test(protoEnumParams.value.toString())) {
```

Comment:
Add a comment here to explain that a value could also be a number so we're checking that here. Otherwise a reader unfamiliar with regexes might not realize what's going on here.

Reply:
Actually I'm not sure why `value` can't be `string | number` as it is in the interface. Is there a reason why it's a string?

I think I can see below that in `encodeValue()` you can then just return the value. But still wonder if the value can be stored as is and converted to other formats as required.

Reply:
In the interface value can be `string | number` because we are giving flexibility to user to either pass enum integer constant or enum-string (like FOLK, ROCK etc in singers example). 

The wire type of ENUM is `string` representation of it's integer constant, so if user pass a numeric value we will assign it as is  and if user passes enum string  we will find it's it's integer constant and then assign it.